docs/proposals/20220927-label-sync-between-machine-and-nodes.md

```diff
@@ -0,0 +1,131 @@
+
+- The `node-restriction.kubernetes.io/` prefix is recommended in the [Kubernetes docs](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#noderestriction) for things such as workload placement; according to https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/ also
+  labels like `example.node-restriction.kubernetes.io/fips=true` falls in this category (prefix =~ domain).
+
+- Cluster API will define its own label prefix, `node.cluster.x-k8s.io/`; for consistency we consider also `example.node.cluster.x-k8s.io/gpu=true` in this category (prefix =~ domain).
+
+#### Synchronization of CAPI Labels
+
+Only the labels that fall under the prefix described above will need to be kept in sync between the Machine and Node objects.
+
+The syncronization process is going to use [server side apply](https://kubernetes.io/docs/reference/using-api/server-side-apply/) in order to ensure that Cluster API will manage only the subset of labels coming from Machine objects and ignore labels applied concurrently by other users/other components.
+
+This requires to define an identity/manager name to be used by CAPI when performing this operation; additionally during implementation we are going to identify and address eventual additional steps required to properly transition existing Nodes to SSA, if required.
+
+The syncronization process will be implemented in the Machine controller, so reconciliation is triggered both when Machine object changes or when the Node changes (the machine controller watches Nodes within the workload cluster).
+
+#### Delay between Node Create and Label Sync
+
+The Node object is first created when kubeadm joins a node to the workload cluster (i.e. kubelet is up and running). There may be a delay (potentially several seconds) before the machine controller kicks in to apply the labels on the Node.
+
+Kubernetes supports both equality and inequality requirements in label selection. In an equality based selection, the user wants to place a workload on node(s) matching a specific label (e.g. Node.Labels contains `my.prefix/foo=bar`). The delay in applying the label on the node, may cause a subsequent delay in the placement of the workload, but this is likely acceptable.
+
+In an inequality based selection, the user wants to place a workload on node(s) that do not contain a specific label (e.g. Node.Labels not contain `my.prefix/foo=bar`). The case is potentially problematic because it relies on the absence of a label and this can occur if the pod scheduler runs during the delay interval.
+
+One way to address this is to use kubelet's `--register-with-taints` flag. Newly minted nodes can be tainted via the taint `cluster.x-k8s.io=label-sync-pending:NoSchedule`. Assuming workloads don't have this specific toleration, then nothing should be scheduled. KubeadmConfigTemplate provides the means to set taints on nodes (see  JoinConfiguration.NodeRegistration.Taints).
```

Comment:
Another option for to work around the inequality problem could be to use it together with equality. If you have a strict inequality requirement then using a selector of the type ~"contains `initialSyncCompleted` and not contains `my.prefix/foo=bar` would work. This feels like it could be tricky for the users and NoSchedule is likely more straight forward. But this would allow all other workloads to just deploy immediately. So just throwing it out there.

Reply:
Interesting idea. So essentially folks would add an additional label everywhere to signal that the sync was done and they can select on that label being there + another one isn't.

@fabriziopandini Wondering if this is good enough to avoid implementation of the entire taint thing.

Considering that probably most of the use cases are selection on equality and are using taints otherwise I think it could be good enough for now (we could have an issue to see if there's demand for a builtin solution vs the trick with labels described here)